### PR TITLE
Make sure we remove a peer on disconnect in gossip

### PR DESCRIPTION
**Pull request by bkchr**

### CODE REVIEW

File: client/network/src/protocol/sync/blocks.rs [hunk 1]
```diff
@@ -165,20 +164,19 @@ impl<B: BlockT> BlockCollection<B> {
...
-					},
-					_ => break,
-				}
+
+		let mut prev = from;
+		for (start, range_data) in &mut self.blocks {
+			match range_data {
+				&mut BlockRangeState::Complete(ref mut blocks) if *start <= prev => {
+						prev = *start + (blocks.len() as u32).into();
+						drained.append(blocks);
```
Comment (tomaka) [inline client/network/src/protocol/sync/blocks.rs]: `append` isn't a very descriptive name, and I had to look up the documentation to check whether this was draining the content of `blocks`. Maybe a small comment mentioning that wouldn't hurt.

Comment (bkchr) [reply client/network/src/protocol/sync/blocks.rs]: Done.

Comment (arkpar) [reply client/network/src/protocol/sync/blocks.rs]: Documentation says `Moves all the elements of other into Self, leaving other empty.` but it does not clarify if "empty" means memory is deallocated.  `mem::replace` makes sure the memory is freed when `blocks` is dropped`

Comment (bkchr) [reply client/network/src/protocol/sync/blocks.rs]: Yeah, but this entry is going to be removed from the hashmap anyway. That will drop the vec.

**Pull request merged**